src/main/kotlin/Network.kt

```diff
@@ -7,24 +7,31 @@ data class NetworkMessage(val message: Message, val deliveryTime: DeliveryTime)
 
 class Network(initialMessages: Map<Address, List<NetworkMessage>> = emptyMap()) {
     var clock = 0
-    private val messages = initialMessages.toMutableMap()
+
+    //TODO Write an ADR
+    // The keys are a Pair because we want to use the Host and Port to get the messages. This is because
+    // Address now has specialised types (Source and Destination) and even tho they have the same values
+    // they are compared differently so the messages were not returned
+    // This is to not break the contract of hashcode and equals
+    private val messages = initialMessages.mapKeys { Pair(it.key.host, it.key.port) }.toMutableMap()
 
     fun get(address: Address): List<Message> {
+        val hostAndPort = Pair(address.host, address.port)
         // get all messages for address with 0 delay and store
-        val (messagesToDeliver, remaining) = messages[address]
+        val (messagesToDeliver, remaining) = messages[hostAndPort]
             ?.partition { (_, deliveryTime) ->
                 deliveryTime <= clock
             } ?: (listOf<NetworkMessage>() to listOf())
 
-        messages[address] = remaining
+        messages[hostAndPort] = remaining
 
         // return store messages
         return messagesToDeliver.map { it.message }
     }
 
     fun add(message: Message, delay: NetworkDelay = 0) {
         val networkMessage = NetworkMessage(message, clock + 1 + delay)
-        messages[message.dest] = messages[message.dest]?.plus(networkMessage) ?: listOf(networkMessage)
+        messages[Pair(message.dest.host, message.dest.port)] = messages[Pair(message.dest.host, message.dest.port)]?.plus(networkMessage) ?: listOf(networkMessage)
```

Comment:
@aidandagnall feel free to propose a different solution also, is there a more idiomatic way to deal with the way I'm creating the pair to be used as keys to retrieve the messages?

Reply:
Sure, you can use `to`, which is maybe more idiomatic, but doesn't really make a big difference

```suggestion
        messages[message.dest.host to message.dest.port] = messages[message.dest.host to message.dest.port]?.plus(networkMessage) ?: listOf(networkMessage)
```

Not a big deal though, might be shorter and nicer to assign it first

```suggestion
        val addressKey = message.dest.host to message.dest.port
        messages[addressKey] = messages[addressKey]?.plus(networkMessage) ?: listOf(networkMessage)
```

Just because it's getting a bit long